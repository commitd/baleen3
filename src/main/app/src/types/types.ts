/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import { Annot8ComponentDescriptor } from './base-types'
import {
  Annot8ComponentInfo,
  Measurement,
  PipelineTemplate,
  SettingsSchema,
  WithDescription,
  WithName,
  ErrorConfiguration,
} from './server-types'

/**
 * Index of metrics measurements
 */
export interface MeasurementContainer {
  [key: string]: Measurement[]
}

/**
 * Holds all the classes that have metrics associated with them, and their associated measurements
 */
export interface MetricsContainer {
  [key: string]: MeasurementContainer
}

export interface SettingsSchemaExtended extends SettingsSchema {
  /**
   * String version of json uiSchema
   *
   */
  uiSchema: string
}

export interface ComponentInfo extends Annot8ComponentInfo {
  name: string
}

/**
 * Map of server components
 */
export interface ComponentMap {
  [key: string]: ComponentInfo
}

/**
 * Map of settings schema
 */
export interface SettingsMap {
  [key: string]: SettingsSchema
}

/**
 * The full set of customisation details for the server.
 */
export interface ServerDetails {
  /**
   * Index by class name of ComponentInfo that represent the processors available on the server that can be used in pipelines.
   */
  processors: ComponentMap
  /**
   * Index by class name of ComponentInfo that represent the sources available on the server that can be used in pipelines.
   */
  sources: ComponentMap
  /**
   * Index by class name of SettingsSchema that are the settings classes of all sources and processors available on the server that can be used in pipelines.
   */
  settings: SettingsMap
  /**
   * Index by class name of ComponentInfo that represent the orderers available on the server that can be used to order pipelines.
   * <p>
   * NB this is currently generated from a strings to simulate a ComponentInfos
   */
  orderers: ComponentMap
  /**
   * Pipeline templates that can be used to start the process of creating a pipeline
   */
  templates: PipelineTemplate[]
}

export interface ComponentMetadata {
  id: string
  descriptor: Annot8ComponentDescriptor
  info: ComponentInfo | undefined
  schema: SettingsSchema | undefined
  valid: boolean
  configured: boolean
}

/**
 * Mapping of Pipeline descriptor to include extra metadata
 *
 */
export interface PipelineHeaderDescriptor extends WithName, WithDescription {}

/**
 * Mapping of Pipeline descriptor to include extra metadata
 *
 */
export interface PipelineViewDescriptor extends PipelineHeaderDescriptor {
  sourceOrder: string[]
  processorOrder: string[]
  components: Record<string, ComponentMetadata>
  errorConfiguration: ErrorConfiguration
}

/**
 * Utility tracking interface to help maintain the UI key state
 */
export interface PipelineEditDescriptor extends PipelineViewDescriptor {
  nameValid: boolean
  orderer: string
  errors: string[]
}

export type ComponentType = 'Processor' | 'Source'

export const NO_SETTINGS: 'io.annot8.api.settings.NoSettings' =
  'io.annot8.api.settings.NoSettings'

export const NO_OP_ORDERER: 'io.annot8.api.pipelines.NoOpOrderer' =
  'io.annot8.api.pipelines.NoOpOrderer'
