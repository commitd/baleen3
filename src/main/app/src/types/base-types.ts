/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
// We remove the generics for the Settings type from these as we never know what they are
// so do not help with any compile time errors

export interface ComponentDescriptor {
  name: string
  settings?: object
}

export interface Annot8ComponentDescriptor {
  [_class: string]: ComponentDescriptor
}

// Types required to complete generated server types
// eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unused-vars-experimental
export interface SourceDescriptor extends Annot8ComponentDescriptor {}
// eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unused-vars-experimental
export interface ProcessorDescriptor extends Annot8ComponentDescriptor {}

/**
 * Interface for describing known metrics, to give better display.
 */
export interface MetricMetadata {
  name?: string
  description?: string
  units?: string
  transform?: (v: number) => number
}

/**
 * Provided stacktrace type to correct optional properties
 */
export interface StackTrace {
  classLoaderName?: string
  className: string
  fileName?: string
  lineNumber: number
  methodName: string
  moduleName?: string
  moduleVersion?: string
  nativeMethod?: boolean
}

/**
 * Provided throwable type to correct optional properties
 */
export interface Throwable {
  cause?: Throwable
  localizedMessage: string
  message: string
  stackTrace: StackTrace[]
  suppressed?: Throwable[]
}
