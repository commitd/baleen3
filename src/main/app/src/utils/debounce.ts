/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
/* eslint-disable @typescript-eslint/no-explicit-any */
export type Method = (...args: any[]) => void

export function debounce<F extends Method>(
  func: F,
  wait = 200
): (this: ThisParameterType<F>, ...args: Parameters<F>) => void {
  let timeoutId: ReturnType<typeof setTimeout> | undefined

  return function (this: ThisParameterType<F>, ...args: Parameters<F>): void {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const context = this

    if (timeoutId !== undefined) {
      clearTimeout(timeoutId)
    }
    timeoutId = setTimeout(() => {
      timeoutId = undefined
      func.apply(context, args)
    }, wait)
  }
}
