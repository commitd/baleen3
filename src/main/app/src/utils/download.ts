/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const downloadJson = (name: string, data: any): void => {
  download(
    name + '.json',
    'data:text/json;charset=utf-8,',
    JSON.stringify(data, null, 2)
  )
}

export const download = (
  name: string,
  mediaType: string,
  data: string
): void => {
  const dataStr = mediaType + encodeURIComponent(data)
  const downloadAnchorNode = document.createElement('a')
  downloadAnchorNode.setAttribute('href', dataStr)
  downloadAnchorNode.setAttribute('download', name)
  document.body.appendChild(downloadAnchorNode) // required for firefox
  downloadAnchorNode.click()
  downloadAnchorNode.remove()
}
