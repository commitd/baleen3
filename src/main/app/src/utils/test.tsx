/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import { StylesProvider } from '@material-ui/styles'
import {
  createHistory,
  createMemorySource,
  History,
  LocationProvider,
} from '@reach/router'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { GenerateId } from 'jss'
import React from 'react'
import { ThemeProvider } from '../components/ThemeProvider/ThemeProvider'

const generateClassName: GenerateId = (rule, styleSheet) => {
  const prefix =
    styleSheet === undefined || styleSheet.options.classNamePrefix === undefined
      ? ''
      : styleSheet.options.classNamePrefix
  return `${prefix}-${rule.key}`
}

const LightTheme: React.FC<{ readonly children?: React.ReactNode }> = ({
  children,
}) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="light">{children}</ThemeProvider>
  </StylesProvider>
)

const DarkTheme: React.FC<{ readonly children?: React.ReactNode }> = ({
  children,
}) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="dark">{children}</ThemeProvider>
  </StylesProvider>
)

export const renderPlain = render

export const renderLight = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: LightTheme, ...options })

export const renderDark = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: DarkTheme, ...options })

export function renderWithRouter(
  ui: Readonly<React.ReactElement>,
  route = '/',
  options?: Omit<RenderOptions, 'queries'>,
  delegate = renderLight
): RenderResult & { history: History } {
  const history = createHistory(createMemorySource(route))
  return {
    ...delegate(
      <LocationProvider history={history}>{ui}</LocationProvider>,
      options
    ),
    history,
  }
}
// re-export everything
export * from '@testing-library/react'
// override render method
export { renderLight as render }
export { userEvent }
