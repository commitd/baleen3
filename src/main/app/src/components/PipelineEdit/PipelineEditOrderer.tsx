/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import React from 'react'
import { Icons } from '@committed/components'
import { ToolbarAction, ToolbarSelect } from '../Header'
import { ComponentMap, NO_OP_ORDERER } from '../../types'

export const PipelineEditOrderer: React.FC<{
  orderers: ComponentMap
  selectedOrderer: string
  onSelect: (selected: string) => void
  onClick: () => void
}> = ({ orderers, selectedOrderer, onSelect, onClick }) => {
  const keys = Object.keys(orderers)
  const options = keys.map((key) => orderers[`${key}`].name)

  //If there are no orderers, or if we only have the NO_OP_ORDERER, then hide this option
  if (
    keys.length === 0 ||
    (keys.length === 1 && orderers[keys[0]].componentClass === NO_OP_ORDERER)
  ) {
    return null
  } else {
    return (
      <>
        <ToolbarSelect
          selectedIndex={keys.indexOf(selectedOrderer)}
          onClick={(selected: number): void => onSelect(keys[Number(selected)])}
          options={options}
          placeholder="Select Orderer"
        />
        <ToolbarAction
          key="apply-orderer"
          disabled={
            selectedOrderer === undefined || selectedOrderer === NO_OP_ORDERER
          }
          icon={<Icons.FormatListNumbered />}
          onClick={onClick}
        >
          Order
        </ToolbarAction>
      </>
    )
  }
}
