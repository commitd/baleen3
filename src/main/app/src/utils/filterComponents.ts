/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import { ComponentInfo, ComponentMap } from '../types'
import { parseSearch, standardize } from './text'

export const filterComponents = (
  map: ComponentMap,
  search: string
): Array<ComponentInfo> => {
  const values = Object.values(map)
  if (search.trim() === '') {
    return values
  }

  const searchTerms = parseSearch(search)

  return values.filter(
    (value) =>
      filterBy('artifact', value, searchTerms) ||
      filterBy('description', value, searchTerms) ||
      filterByName(value, searchTerms) ||
      filterByTags(value, searchTerms)
  )
}

function filterBy(
  key: keyof Omit<ComponentInfo, 'tags'>,
  value: ComponentInfo,
  terms: string[]
): boolean {
  // false positive security/detect-object-injection
  // eslint-disable-next-line
  const property = value[key]
  if (property === undefined) {
    return false
  }
  return terms.some((term) => standardize(property).includes(term))
}

function filterByName(value: ComponentInfo, terms: string[]): boolean {
  const property = standardize(value.name)
  return terms.some((term) => property.includes(term))
}

function filterByTags(value: ComponentInfo, terms: string[]): boolean {
  const tags = value.tags?.map((tag) => tag.toLowerCase())
  return terms.some((term) => tags?.includes(term))
}
