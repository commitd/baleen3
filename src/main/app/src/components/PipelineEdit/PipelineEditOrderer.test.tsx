/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import React from 'react'
import { PipelineEditOrderer } from './PipelineEditOrderer'
import { NO_OP_ORDERER } from '../../types'
import { exampleOrderer } from '../../types/examples'
import { render, renderDark } from '../../utils/test'

jest.mock('../../containers/PipelineMetricsContainer', () => {
  const FakePageContainer = jest.fn(() => null)
  return { PipelineMetricsContainer: FakePageContainer }
})

it('renders without error, with no orderers', () => {
  const { asFragment } = render(
    <PipelineEditOrderer
      orderers={{}}
      selectedOrderer={NO_OP_ORDERER}
      onSelect={jest.fn()}
      onClick={jest.fn()}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders dark without error, with no orderers', () => {
  const { asFragment } = renderDark(
    <PipelineEditOrderer
      orderers={{}}
      selectedOrderer={NO_OP_ORDERER}
      onSelect={jest.fn()}
      onClick={jest.fn()}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders without error, with No-Op orderer', () => {
  const orderers = {
    NO_OP_ORDERER: {
      name: 'No Op Orderer',
      componentClass: NO_OP_ORDERER,
      artifact: 'orderer',
    },
  }

  const { asFragment } = render(
    <PipelineEditOrderer
      orderers={orderers}
      selectedOrderer={NO_OP_ORDERER}
      onSelect={jest.fn()}
      onClick={jest.fn()}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders dark without error, with No-Op orderer', () => {
  const orderers = {
    NO_OP_ORDERER: {
      name: 'No Op Orderer',
      componentClass: NO_OP_ORDERER,
      artifact: 'orderer',
    },
  }

  const { asFragment } = renderDark(
    <PipelineEditOrderer
      orderers={orderers}
      selectedOrderer={NO_OP_ORDERER}
      onSelect={jest.fn()}
      onClick={jest.fn()}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders without error, with orderers', () => {
  const orderers = {
    NO_OP_ORDERER: {
      name: 'No Op Orderer',
      componentClass: NO_OP_ORDERER,
      artifact: 'orderer',
    },
    EXAMPLE_ORDERER: exampleOrderer,
  }

  const { asFragment } = render(
    <PipelineEditOrderer
      orderers={orderers}
      selectedOrderer="EXAMPLE_ORDERER"
      onSelect={jest.fn()}
      onClick={jest.fn()}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders dark without error, with orderers', () => {
  const orderers = {
    NO_OP_ORDERER: {
      name: 'No Op Orderer',
      componentClass: NO_OP_ORDERER,
      artifact: 'orderer',
    },
    EXAMPLE_ORDERER: exampleOrderer,
  }

  const { asFragment } = renderDark(
    <PipelineEditOrderer
      orderers={orderers}
      selectedOrderer="EXAMPLE_ORDERER"
      onSelect={jest.fn()}
      onClick={jest.fn()}
    />
  )
  expect(asFragment()).toMatchSnapshot()
})
