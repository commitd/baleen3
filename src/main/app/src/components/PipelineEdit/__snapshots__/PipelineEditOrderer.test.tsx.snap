// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders dark without error, with No-Op orderer 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`renders dark without error, with no orderers 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`renders dark without error, with orderers 1`] = `
<DocumentFragment>
  <div>
    <button
      aria-haspopup="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text WithNoOpener-root WithNoOpener-root MuiButton-textSizeSmall MuiButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Example Orderer
        <span
          class="MuiButton-endIcon MuiButton-iconSizeSmall"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5H7z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text WithNoOpener-root WithNoOpener-root MuiButton-textSizeSmall MuiButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeSmall"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2 17h2v.5H3v1h1v.5H2v1h3v-4H2v1zm1-9h1V4H2v1h1v3zm-1 3h1.8L2 13.1v.9h3v-1H3.2L5 10.9V10H2v1zm5-6v2h14V5H7zm0 14h14v-2H7v2zm0-6h14v-2H7v2z"
            />
          </svg>
        </span>
        Order
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`renders without error, with No-Op orderer 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`renders without error, with no orderers 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`renders without error, with orderers 1`] = `
<DocumentFragment>
  <div>
    <button
      aria-haspopup="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text WithNoOpener-root WithNoOpener-root MuiButton-textSizeSmall MuiButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Example Orderer
        <span
          class="MuiButton-endIcon MuiButton-iconSizeSmall"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5H7z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text WithNoOpener-root WithNoOpener-root MuiButton-textSizeSmall MuiButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeSmall"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2 17h2v.5H3v1h1v.5H2v1h3v-4H2v1zm1-9h1V4H2v1h1v3zm-1 3h1.8L2 13.1v.9h3v-1H3.2L5 10.9V10H2v1zm5-6v2h14V5H7zm0 14h14v-2H7v2zm0-6h14v-2H7v2z"
            />
          </svg>
        </span>
        Order
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
